Parallelize autograd backwards

colesbury opened the pull request:
Parallelize autograd.backward()

Moves convolution and batch normalization functions into C++ to avoid holding onto the GIL.

colesbury committed:
Parallelize backwards

colesbury committed:
Improve broadcast/reduce performance by coalescing tensors

albanD commented on torch/csrc/autograd/python_engine.cpp:
```diff
@@ -3,13 +3,33 @@
+struct PythonEngine : public Engine {
+  virtual void thread_main(ReadyQueue& queue) override {
+    AutoGIL gil;
+    AutoNoGIL no_gil;
```
Is the point of these two lines that, this new threads needs to have a thread state associated because some Python function may be called inside (somewhere during the backward), but we actually don't want to hold the GIL right now and the GIL will be locked just before calling the python function later?

colesbury replied:
Yes. From the Python source code it looked like the first and last calls to `PyGILState_Ensure` and  `PyGILState_Release` are more expensive since they create and delete `PyThreadState` objects. Without this, internal calls will create and delete `PyThreadState` objects instead of just acquiring and releasing the lock.  It would still be correct without these calls, but may be more expensive.

I'm not entirely sure about the cost -- I haven't measured it without these lines.

apaszke replied:
Can you please add a comment with all that? It looks a weird when you don't have the context.

albanD commented on torch/csrc/autograd/python_hook.cpp:
```diff
@@ -0,0 +1,42 @@
+namespace torch { namespace autograd {
+
+PyGradHook::PyGradHook(PyObject* dict) : dict(dict) {
+  Py_INCREF(dict);
```
Is there any reason why you get the GIL for every operation on `PyGradHook` except for creation?

colesbury replied:
The destructor and `operator()` are virtual overloads. They can be called from code that doesn't even know about Python so we have to acquire the GIL.

The constructor (by nature of C++) is not virtual. Since it can only be constructed by name (`PyGradHook(obj)`), it follows Python conventions for object ownership and assumes the calling code holds onto the GIL.

apaszke replied:
Because it's only called from [here](https://github.com/pytorch/pytorch/pull/881/files#diff-1ceb14cfe86ae521dee960ed8fc3daeeR79) and the thread has the GIL in that context.

albanD replied:
Ok thanks

apaszke commented on torch/csrc/autograd/function.cpp:
```diff
@@ -28,4 +28,24 @@ auto Function::flags(const variable_list& inputs) -> FunctionFlags {
+  variable_list outputs(inputs.size());
+  for (size_t i = 0; i < inputs.size(); ++i) {
+    if (hooks[i]) {
+      outputs[i] = (*hooks[i])(inputs[i]);
```
Wouldn't it be better to have either a single hook that gets all outputs (and maybe calls other hooks with some of them), or a number of hooks from all outputs to all outputs?

apaszke commented on torch/csrc/autograd/functions/batch_normalization.cpp:
```diff
@@ -111,7 +117,7 @@ auto BatchNormBackward::apply(const variable_list& grad_outputs) -> variable_lis
+  if (needs_input_grad(2) || use_cudnn) {
     grad_bias = bias->newTensor();
     grad_bias->resizeAs(*bias);
     grad_bias->zero();
```
It seems that there's another `zero()` call 2 lines below. Can you fix that please?

apaszke commented on torch/csrc/autograd/functions/init.cpp:
```diff
@@ -48,9 +64,18 @@ static void addClass(PyObject* module, PyTypeObject& type, const char* name)
   addClass<BatchNormForward, BatchNormCtor>(module, BatchNormClass, "BatchNorm");
   addClass<BatchNormBackward, NoCtor>(module, BatchNormBackwardClass, "BatchNormBackward");
+
+  static PyTypeObject ConvClass, ConvBackwardClass;
```
Can you move the BatchNorm class variables into this function too?

apaszke commented on torch/csrc/utils/tuple_parser.cpp:
```diff
@@ -0,0 +1,85 @@
+   }
+ }
+
+auto TupleParser::parse(bool& x) -> void {
```
It would be nicer to be able to pass multiple arguments into the call, as you can do to `_appendData` in THD. The only downside is that it instantiates a new template for each configuration of arguments, but it's more readable.

apaszke commented on torch/csrc/utils/tuple_parser.cpp:
```diff
@@ -0,0 +1,85 @@
+  msg += " must be ";
+  msg += expected;
+  msg += ", not ";
+  msg += Py_TYPE(PyTuple_GET_ITEM(args, idx - 1))->tp_name;
```
It's better to use `THPUtils_typename`.

apaszke commented on torch/nn/functional.py:
```diff
@@ -3,9 +3,10 @@
 from .modules.utils import _single, _pair, _triple
+
 # Convolutions
+Conv = torch._C._functions.Conv
```
Why not ConvNd? 😞

apaszke commented on torch/csrc/autograd/python_cpp_function.cpp:
```diff
@@ -79,12 +94,18 @@ int TensorConverter(PyObject* obj, std::unique_ptr<thpp::Tensor>* address)
 }
 
+static struct PyMethodDef THPCppFunction_methods[] = {
+  {(char*)"_register_hook_dict", (PyCFunction)THPCppFunction_register_hook_dict, METH_O, NULL},
```
So did the C functions support hooks on Variables before that?

apaszke commented on torch/csrc/autograd/python_hook.cpp:
```diff
@@ -0,0 +1,42 @@
+    if (!res) throw python_error();
+    if (res == Py_None) continue;
+    if (!PyObject_IsInstance(res.get(), THPVariableClass)) {
+      PyErr_Format(PyExc_TypeError, "expected Variable, but hook returned '%s'",
```
What happened to all of my other checks :(

apaszke commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -0,0 +1,56 @@
+    return (long)value;
+  }
+#if PY_MAJOR_VERSION == 2
+  if (PyInt_Check(obj)) {
```
There are `PyLong`s in Python 2 too.

apaszke commented on torch/csrc/utils/python_numbers.h:
```diff
@@ -0,0 +1,56 @@
+    if (overflow != 0) {
+      throw std::runtime_error("Overflow when unpacking double");
+    }
+    return (double)value;
```
If we're checking for overflows, then I'd add a check to make sure that it's smaller than 1e53 too.

apaszke commented on torch/nn/parallel/_functions.py:
```diff
@@ -9,13 +8,19 @@ def __init__(self, target_gpus):
+        return tuple([t for tensors in outputs for t in tensors])
+
+    def backward(self, *grad_outputs):
+        grad_outputs = [grad_outputs[i:i+self.num_inputs]
```
Where is `self.num_inputs` assigned? I can't see that.

colesbury committed:
Implement 'pre' and 'post' hooks at the C++ autograd level

apaszke commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -1,15 +1,198 @@
+
+    std::lock_guard<std::mutex> lock(task.base->mutex);
+    if (auto var = dynamic_cast<Variable*>(prev_fn.get())) {
+      if (!grad_input) {
```
Ok, can you leave a `FIXME` or `NOTE` there?

apaszke commented on torch/csrc/autograd/functions/convolution.cpp:
```diff
@@ -0,0 +1,503 @@
+
+  bool use_cudnn = false;
+#ifdef WITH_CUDNN
+  // TODO: cuDNN version
```
?

colesbury replied:
Newer versions of cuDNN support dilated convolution. I'll fix this.

apaszke commented on torch/csrc/autograd/functions/convolution.cpp:
```diff
@@ -0,0 +1,503 @@
+
+  bool use_cudnn = false;
+#ifdef WITH_CUDNN
+  // TODO: cuDNN version
```
As above. It seems that it's implemented.

apaszke commented on torch/nn/parallel/replicate.py:
```diff
@@ -1,42 +1,73 @@
+    buffer_copies = comm.broadcast_coalesced(buffers, devices)
+
+    modules = list(network.modules())
+    module_copies = [[] for device in devices]
```
I think it would be better to store the transpose of that, because you always have to reverse the indexing in the main loop, but it's not very important. All of `for j in range(num_replicas):` could be replaced with `for replica in module_copies[i]:`.

apaszke commented on torch/nn/parallel/replicate.py:
```diff
@@ -1,42 +1,73 @@
+                    replica = module_copies[j][i]
+                    replica._buffers[key] = buffer_copies[j][buffer_idx]
+
+    return [module_copies[j][0] for j in range(num_replicas)]
```
Why does it return only the first module for each replica?

colesbury replied:
The first module is the outermost module (i.e. `list(network.modules())[0] == network`), which is the same module that the user passed as an argument to replicate.

This is the same behavior that replicate had before:
```
  replicate(network, devices) -> replicas
```
where replicas is a list of copies of the module `network` (i.e.`len(replicas) == len(devices)`)

colesbury commented:
I've rebase on top of master and moved the hook implementation to C++. You may want to look at:

- engine.cpp (rebase)
- The commit "Implement 'pre' and 'post' hooks at the C++ autograd level"

albanD commented on torch/lib/THPP/Type.hpp:
```diff
@@ -43,4 +43,26 @@ inline bool isInteger(Type t) {
+    case Type::LONG_STORAGE: return "LongStorage";
+    case Type::TENSOR: return "Tensor";
+    case Type::STORAGE: return "Storage";
+    default: return "";
```
Is it ever possible to reach the default here?
If so, do we want and "<Unknow Type>" message to avoid having empty string instead of types in error messages?

apaszke replied:
It is, if someone extends the enum, but forgets to add a case here.

apaszke commented on torch/csrc/autograd/engine.cpp:
```diff
@@ -82,106 +288,75 @@ auto Engine::backward(const variable_list& variables,
-      }
-    }
+  // check for an exception while running backwards
+  if (backward_task.has_error.load()) {
```
nit: `std::atomic` always puts a memory fece around reads, so there's no need to call `load()` explicitly. But we can leave that

apaszke commented on torch/csrc/autograd/functions/batch_normalization.h:
```diff
@@ -8,65 +8,47 @@
-
-  std::unique_ptr<thpp::Tensor> running_mean;
-  std::unique_ptr<thpp::Tensor> running_var;
+struct BatchNormParams {
```
nit: I think state or context would be better than params

colesbury commented on torch/csrc/autograd/python_function.h:
```diff
@@ -34,7 +34,6 @@ struct THPFunction {
 
     PyObject *needs_input_grad;
     PyObject *backward_hooks;
```
We expose `fn._backward_hooks` as a field. We could either reach-in and grab the PyObject* if post_hooks has a `PyFunctionPostHook` or leave this here.

apaszke replied:
No, no. The only solution would be to port `register_hook` to C (that's the only place where we use the attribute), but I think we can leave it as is for now if you want.

colesbury commented on torch/csrc/autograd/python_hook.cpp:
```diff
@@ -1,42 +1,189 @@
+  if (original.sizes() != result.sizes()) {
+    std::stringstream ss;
+    auto name = hook_name(hook);
+    ss << "backward hook '" << name << "' has changed the size of grad_input";
```
Probably... these are the same error messages that were in python_function.cpp previously.

colesbury commented on torch/csrc/autograd/python_variable.cpp:
```diff
@@ -292,10 +294,9 @@ int THPVariable_set_backwards_hooks(THPVariable *self, PyObject *obj)
-  } else {
-    self->cdata->backward_hook.reset();
+    self->cdata->pre_hooks.emplace_back(new PyFunctionPreHook(obj, 0));
   }
```
It will remove the old one (`self->cdata->pre_hooks.clear()`).

apaszke replied:
it will remove them, but only from the Variable. The content of the field from the Variable doesn't matter, the important field is that vector inside the creator that will still remember the old value.

colesbury replied:
This code doesn't touch the creator

colesbury commented on torch/csrc/Exceptions.h:
```diff
@@ -24,6 +22,45 @@ class python_error : public std::exception {};
+    Py_XINCREF(type);
+    Py_XINCREF(value);
+    Py_XINCREF(traceback);
+    PyErr_Restore(type, value, traceback);
```
Then the python_error can only be restored once. That makes the API more complicated.

colesbury commented:
Add `Py_VISIT` calls and a test for cycles

apaszke commented on torch/csrc/autograd/python_function.cpp:
```diff
@@ -125,14 +131,13 @@ static int THPFunction_clear(THPFunction *self)
   self->output_info = NULL;
   delete output_info;
 
+  self->cdata.pre_hooks.clear();
```
Doesn't this need something like `Py_CLEAR`? I'm afraid that in the middle of the `clear()` call, the GC might kick in, and will call traverse. I think it would be safer to swap with a local empty vector (no need to call `clear()`, because it will be destructed just a moment later).

colesbury commented:
(rebased)

colesbury committed:
Fix bug where adding a hook could replace an existing hook.

We were keying hooks by RemovableHandle id. However, we don't hold onto
handles and ids of dead objects can be reused. This replaces id(handle)
with a global counter.
